dodo/pgp_util.py

```diff
@@ -80,9 +79,12 @@ def encrypt(msg: email.message.EmailMessage) -> email.message.EmailMessage:
             val for key, val in msg.items() if key in ['From', 'To', 'Cc']
         ])
     ]
-    recipients_keys = [k['fingerprint'] for k in Gpg.list_keys()
-                       if any(re.search(addr, u) for u in k['uids']
-                              for addr in recipients)]
+    # set strict=False for parseaddr, otherwise keys with a uid such as
+    # "mail@example.org <mail@example.org>" are skipped. These uids are considered
+    # malformed, but they occur.
+    recipients_keys = [key['fingerprint'] for key in Gpg.list_keys()
+                       if any(addr == email.utils.parseaddr(uid, strict=False)[1]
+                           for uid in key['uids'] for addr in recipients)]
```

Comment:
I feel like this gets quite cramped with `[... for ... in ... if any(... for ... in ... for ... in ...)]`. That's a lot going on in one list comprehension. What do you think about a (untested):

```python
def find_gpg_keys(recipients: list[str]) -> list[str]:
    keys = []
    for key in Gpg.list_keys():
        for uid in key["uid"]:
            address = utils.strip_email_address(uid, strict=False)
            if address in recipients:
                keys.append(key["fingerprint"])
```

or maybe

```python
def find_gpg_keys(recipients: list[str]) -> list[str]:
    keys = []
    for key in Gpg.list_keys():
        addresses = {utils.strip_email_address(uid, strict=False) for uid in key["uids"]}
        if addresses & recipients:
            keys.append(key["fingerprint"])
```

Reply:
I removed the strict parameter and used the `util.strip_email_address` instead. I will address the case where `gnupg.list_keys` emits malformed email addresses later, together with your suggestions above. Note that the behavior of `util.strip_email_address` changes in python 3.12.6 but I think this will not have a big impact on dodo because, at first sight,  notmuch transforms mail addresses such as `mail@example.com <mail@example.com>` to  `"mail@example.com" <mail@example.com>` which are parsed correctly by the new `email.utils.parseaddr` (and also by the golang validation mentioned on [issue #102988](https://github.com/python/cpython/issues/102988))